From 5c73c35e055a2fd609fa079db3aba171db37d1cb Mon Sep 17 00:00:00 2001
From: Jarod Wilson <jarod@redhat.com>
Date: Wed, 7 Jan 2015 15:40:00 -0500
Subject: [crypto] drbg: panic on continuous self test error

Message-id: <1420645200-2574-1-git-send-email-jarod@redhat.com>
Patchwork-id: 102515
O-Subject: [RHEL7 PATCH] crypto: drbg - panic on continuous self test error
Bugzilla: 1179496
RH-Acked-by: Nikolay Aleksandrov <nikolay@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Aristeu Rozanski <aris@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1179496

Trivial backport required for FIPS compliance.

From Herbert's cryptodev-2.6.git tree.

commit 905b42e559fa4952569b3444bc6c054c0103e5a0
Author: Stephan Mueller <smueller@chronox.de>
Date:   Fri Dec 5 22:40:21 2014 +0100

    crypto: drbg - panic on continuous self test error

    This patch adds a panic if the FIPS 140-2 self test error failed.
    Note, that entire code is only executed with fips_enabled (i.e. when the
    kernel is booted with fips=1. It is therefore not executed for 99.9% of
    all user base.

    As mathematically such failure cannot occur, this panic should never be
    triggered. But to comply with NISTs current requirements, an endless
    loop must be replaced with the panic.

    When the new version of FIPS 140 will be released, this entire
    continuous self test function will be ripped out as it will not be
    needed any more.

    This patch is functionally equivalent as implemented in ansi_cprng.c and drivers/char/random.c.

    Signed-off-by: Stephan Mueller <smueller@chronox.de>
    Signed-off-by: Herbert Xu <herbert@gondor.apana.org.au>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/crypto/drbg.c b/crypto/drbg.c
index 68eee2a..9999b46 100644
--- a/crypto/drbg.c
+++ b/crypto/drbg.c
@@ -222,15 +222,6 @@ static inline unsigned short drbg_sec_strength(drbg_flag_t flags)
  * function. Thus, the function implicitly knows the size of the
  * buffer.
  *
- * The FIPS test can be called in an endless loop until it returns
- * true. Although the code looks like a potential for a deadlock, it
- * is not the case, because returning a false cannot mathematically
- * occur (except once when a reseed took place and the updated state
- * would is now set up such that the generation of new value returns
- * an identical one -- this is most unlikely and would happen only once).
- * Thus, if this function repeatedly returns false and thus would cause
- * a deadlock, the integrity of the entire kernel is lost.
- *
  * @drbg DRBG handle
  * @buf output buffer of random data to be checked
  *
@@ -257,6 +248,8 @@ static bool drbg_fips_continuous_test(struct drbg_state *drbg,
   return false;
  }
  ret = memcmp(drbg->prev, buf, drbg_blocklen(drbg));
+ if (!ret)
+  panic("DRBG continuous self test failed\n");
  memcpy(drbg->prev, buf, drbg_blocklen(drbg));
  /* the test shall pass when the two compared values are not equal */
  return ret != 0;
-- 
1.7.1